### PR TITLE
Adds a status page to Library and Playlists pages.

### LINKED ISSUE #305: [Feature Request] Improve bording experience
New spotify users will most likely have there library and playlist empty, which may users quickly switch back to the official spotify app. I think it would be nice to have a special case for empty categories with text saying why the list is empty and a suggestion how they can fill the view.  This can be done with [AdwStatusPage](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.0.0-alpha.2/class.StatusPage.html)

In GNOME we generally have something like this (this is Fractal) 
image

### PR DESCRIPTION
**Pull request by Diegovsky**
This PR will hopefully fix #305. I'm new to the project and am still getting used to the codebase so I might have edited the wrong files.

## Unresolved stuff
 - What should go in Library's
   - [ ] title
   - [ ] description
   - [x] icon
 - What should go in Playlists'
   - [ ] title
   - [ ] description
   - [x] icon

### CODE REVIEW

**Comment by Diegovsky** [comment]
Here's how it looks as of f2bc8a5 
Library
image

**Comment by xou816** [comment]
Hi! Thanks for taking a look at this! :)

Wording: maybe drop the "it seems..." to make them more consistent/easier to translate?
"Consider creating one" -> I haven't implemented this yet, let's avoid mentioning this for now :D 

As for the icon, I don't really know -- I guess the emblem-music-symbolic icon could do!

You forgot to commit the UI files however

**Comment by Diegovsky** [comment]
Great. I'll change some of the wording and add the icons. 

> You forgot to commit the UI files however

I dindn't edit them. AFAIK they are descriptions of how a widget should look like, right? The status pages are shown dynamically. I'm not sure what needs to be changed in that case,

**Comment by xou816** [comment]
> Great. I'll change some of the wording and add the icons.

Thanks! When we're done with choosing the wordings I'll add them to POEditor.

> I dindn't edit them. AFAIK they are descriptions of how a widget should look like, right? The status pages are shown dynamically. I'm not sure what needs to be changed in that case,

Oh right, missed how you implemented this. It works I guess, but for instance if you were to remove albums from your saved albums so as to go back to the empty state, it wouldn't hide the flowbox and use the status page instead. I think it'd be better to have an overlay, with the status page and the flowbox, and then switch the visible widget in on_event. Hopefully that makes sense!

(oh and it looks like the CI complains about formatting)

**Comment by Diegovsky** [comment]
>  Oh right, missed how you implemented this. It works I guess, but for instance if you were to remove albums from your saved albums so as to go back to the empty state, it wouldn't hide the flowbox and use the status page instead.

In that case I will probably end up doing what you said instead :)

**Comment by Diegovsky** [comment]
I'll open another PR once I complete the .ui file changes. 

>I think it'd be better to have an overlay, with the status page and the flowbox, and then switch the visible widget in on_event. 

What would be a good way of knowing if the user has no albums or playlists?

**Comment by xou816** [comment]
You can reopen this one if you want to, that's fine!

> What would be a good way of knowing if the user has no albums or playlists?

A quick overview of the project's architecture:
- components have read only access to a state object (the state of the whole app)
- components that implement EventListener are notified of changes to the state

Therefore you could wait for specific events related to the state of these screens, and read the current state at that point.

**Comment by Diegovsky** [comment]
> You can reopen this one if you want to, that's fine!

I reset my branch to master and it accidentally closed so I went with it lol
> A quick overview of the project's architecture: ...

Neat, seems like I might finish today :)

**Comment by Diegovsky** [comment]
I edited the .ui files and added a GtkOverlay as you mentioned and it was surprisingly clean :)

Since I can't remove all my playlists, I can't test this right now unfortunately

**Comment by xou816** [comment]
Sorry forgot about this! I'll have a look tomorrow!

**Comment by Diegovsky** [comment]
Don't sweat it, I'm on summer break anyways :)

File: src/app/components/library/library.rs [hunk 1]
```diff
@@ -131,6 +134,13 @@ impl EventListener for Library {
             AppEvent::LoginEvent(LoginEvent::LoginCompleted(_)) => {
                 let _ = self.model.refresh_saved_albums();
             }
+            AppEvent::BrowserEvent(BrowserEvent::LibraryUpdated) => {
+                if let Some(list) = self.model.get_list_store() {
```
Comment (xou816) [inline src/app/components/library/library.rs:138]: (minor) you could add a has_albums method to the model

File: src/app/components/library/library.rs [hunk 2]
```diff
@@ -131,6 +134,13 @@ impl EventListener for Library {
...
+                    imp::LibraryWidget::from_instance(&self.widget)
+                        .status_page
```
Comment (xou816) [inline src/app/components/library/library.rs:140]: (minor) there could be a getter on the widget for consistency

File: src/app/components/library/library.ui [hunk 3]
```diff
@@ -9,14 +9,26 @@
...
+                <property name="margin-top">8</property>
+                <property name="margin-bottom">8</property>
+                <property name="min-children-per-line">1</property>
+                <property name="selection-mode">none</property>
+                <property name="activate-on-single-click">0</property>
+              </object>
+            </child>
+            <child type="overlay">
+              <object class="AdwStatusPage" id="status_page">
+                <property name="title" translatable="yes" comments="A title that is shown when the user has not saved any albums.">You have no saved albums</property>
```
Comment (xou816) [inline src/app/components/library/library.ui:26]: perfect with comments and everything :) I'll add the strings on POEditor for translators. do you mind just ending with a period?

File: src/app/components/saved_playlists/saved_playlists.rs [hunk 4]
```diff
@@ -134,6 +136,13 @@ impl EventListener for SavedPlaylists {
             AppEvent::LoginEvent(LoginEvent::LoginCompleted(_)) => {
                 let _ = self.model.refresh_saved_playlists();
             }
+            AppEvent::BrowserEvent(BrowserEvent::SavedPlaylistsUpdated) => {
+                if let Some(list) = self.model.get_list_store() {
```
Comment (xou816) [inline src/app/components/saved_playlists/saved_playlists.rs:140]: (minor) (same 2 comments as above)

File: src/app/components/saved_playlists/saved_playlists.ui [hunk 5]
```diff
@@ -9,14 +9,26 @@
...
+                <property name="margin-top">8</property>
+                <property name="margin-bottom">8</property>
+                <property name="min-children-per-line">1</property>
+                <property name="selection-mode">none</property>
+                <property name="activate-on-single-click">0</property>
+              </object>
+            </child>
+            <child type="overlay">
+              <object class="AdwStatusPage" id="status_page">
+                <property name="title" translatable="yes" comments="A title that is shown when the user has not saved any playlists.">You have no saved playlists</property>
```
Comment (xou816) [inline src/app/components/saved_playlists/saved_playlists.ui:26]: same here, great job, just missing a period :)

**Review by xou816 — COMMENTED** [review]
great job overall, just minor nitpicks :p thank you!

**Comment by xou816** [comment]
(the diff is weird, maybe you need to update your master branch, but its fine as there was only one commit to review.)

**Comment by xou816** [comment]
oh wait I forgot! you need to update the terms with `ninja spot-pot -C target`

File: src/app/components/saved_playlists/saved_playlists.ui [hunk 6]
```diff
@@ -9,14 +9,26 @@
...
+                <property name="description" translatable="yes" comments="A description of what happens when the user has saved playlists.">Your playlists will be shown here.</property>
+                <property name="icon-name">emblem-music-symbolic</property>
+                <property name="visible">false</property>
```
Comment (xou816) [inline src/app/components/saved_playlists/saved_playlists.ui:29]: it might need to be true initially (first login for instance)

**Review by xou816 — CHANGES_REQUESTED** [review]
a little something I found while testing

File: src/app/components/library/library.ui (hunk 3 above)
Comment (Diegovsky) [reply src/app/components/library/library.ui:26]: Ok

File: src/app/components/saved_playlists/saved_playlists.ui (hunk 6 above)
Comment (Diegovsky) [reply src/app/components/saved_playlists/saved_playlists.ui:29]: Indeed, it makes more sense

**Comment by Diegovsky** [comment]
About the diff being weird: I'm more or less intermediate with git. I tried pulling your changes from master to make it easy to merge but it probably got worse lol

I'll ask a friend to help me fix this weird history before merge

**Comment by xou816** [comment]
no worries I think it's fine on your branch, which is all that really matters, but maybe not on the master branch of your fork!

File: src/app/components/library/library.rs (hunk 1 above)
Comment (xou816) [reply src/app/components/library/library.rs:141]: depends how you see it I guess, but what you care about locally is whether there are albums or not (and this avoids having to deal with the option here) but im fine with either :)

**Comment by xou816** [comment]
I'll have a look to see why the pipeline fails, this is not on you I believe

Oh and sorry but you have to run the spot-pot thing again since you added the periods :p

**Comment by xou816** [comment]
That's on me, I should definitely stick to a specific revision of libadwaita. I'll fix it later!

**Comment by xou816** [comment]
Hi! you should be able to rebase on master, CI should be good now :)

File: src/app/components/library/library.rs (hunk 1 above)
Comment (Diegovsky) [reply src/app/components/library/library.rs:141]: That's true. I overlooked it and refreshing the albums is done elsewhere so a `has_albums` method is actually the better for this case :)

**Comment by Diegovsky** [comment]
Alrighty! I'll just add the `has_` methods, rerun spot-pot, and edit my messages. I am rebasing in a few moments :)

**Review by xou816 — APPROVED** [review]
perfect!

**Comment by xou816** [comment]
There is still that weird diff, I don´t know why, have you been able to figure out whats happening? it should be fine to merge but still weird

**Comment by Diegovsky** [comment]
> There is still that weird diff, I don´t know why, have you been able to figure out whats happening? it should be fine to merge but still weird

Probably because I pulled from master and it messed the history. I won't do it again from now on lol

Also, it was a pleasure to work on this with you! Thanks for the kindness and patience :D

**Comment by xou816** [comment]
It's been a pleasure as well, thanks for your contribution! :)

**Pull request merged**